REF: add inline value refactoring

Kobzol opened the pull request:
This PR adds support for inlining variables and constants. Right now the original variable is inlined and removed. If it does not have any usages, an error is reported. In Kotlin, there is an additional dialog that asks if the original expression should be kept or not, should I also include it here?

I wasn't sure which test infrastructure to use as it's not a classic refactoring, so I added some manual test functions, but it's not a lot of code.

~Functions could also be inlined, should that be handled inside this refactoring?~

Fixes: https://github.com/intellij-rust/intellij-rust/issues/1095

vlad20012 commented on src/main/kotlin/org/rust/ide/refactoring/inlineValue/InlineValueHandler.kt:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    override fun canInlineElement(element: PsiElement): Boolean {
+        if (element.ancestorOrSelf<RsConstant>() != null) return true
```
I wonder why Java's implementations like `InlineLocalHandler`, `InlineConstantFieldHandler`, `InlineMethodHandler` just check the type of `element` without checking ancestors

Kobzol replied:
If I just check for the type directly, it is not available for the whole constant/let declaration, which seems unintuitive for me:
```rust
const CONST: u32 = 5/*caret*/; // not available
const CONST: u32/*caret*/ = 5; // not available
const CONST:/*caret*/ u32 = 5; // not available
const CONST/*caret*/: u32 = 5; // available
/*caret*/const CONST: u32 = 5; // not available
```

Kobzol replied:
Actually to be on par with Java, we also need to check if the element is a reference to a constant/let decl, for this to work:
```rust
const CONST: u32 = 1;
let a = CONST/*caret*/ + 2;
```

vlad20012 replied:
Clear! But this is exactly how it works in Java. The refactoring is available only when the caret is placed near a constant name or a reference to it.

Kobzol replied:
Okay, so should I change it to behave the same as Java/Kotlin?

vlad20012 replied:
> Actually to be on par with Java, we also need to check if the element is a reference to a constant/let decl, for this to work

This works automagically right now =)

> so should I change it to behave the same as Java/Kotlin?

I think, yes

Kobzol commented:
I rewrote the refactoring to be in line (no pun intended) with the InlineFunction refactoring. The UI dialog was mostly the same, so I refactored it out, along with some other things and tests in InlineFunction. There is now also support for testing the various inlining modes (only this/all references/all references and keep original) for InlineValue. Refactoring from references is now also supported.

Kobzol commented:
Is there anything left to do here? After this PR I'll add support for inlining of parameters in the inline function refactoring.

LoganDark commented:
Really missing this feature right now. It's much better than manual copy and paste

Kobzol commented:
@vlad20012 ping

AqlaSolutions commented:
Yep, I need this too

LoganDark commented:
I'd like to know why this PR is still sitting around, without any comments from the maintainers on why it's not being merged.

@vlad20012 Replying with an 👀 emoji just serves to tell us that you see this PR and you're intentionally ignoring it.

Kobzol commented:
Take it easy, the maintainers are doing a lot of work to keep the plugin stable (fixes, performance improvements, internal refactorings etc.) and adding new features might not always be a top priority. I myself got almost 100 PRs merged in the last 6 months, so there's no problem with the maintainers accepting PRs :) I'm sure the review on this PR will continue in a reasonable timeframe.

LoganDark commented:
They've looked at it multiple times, all they have to do is hit the merge button. I personally don't see the reasoning for saying that they're doing other things and this might not be a top priority; it's priority enough for them to look at it, and react to comments, and review it and ask for changes.

vlad20012 reviewed (approved):
I'm very sorry for the long response(
I also apologize to the people who were waiting for this feature.

Thanks!
bors r+
